src/script_compiler.rs

```diff
@@ -29,15 +37,87 @@ extern "C" {
 #[derive(Debug)]
 pub struct Source([usize; 8]);
 
+/// Compilation data that the embedder can cache and pass back to speed up future
+/// compilations. The data is produced if the CompilerOptions passed to the compilation
+/// functions in ScriptCompiler contains produce_data_to_cache = true. The data to cache
+/// can then can be retrieved from UnboundScript.
+#[repr(C)]
+#[derive(Debug)]
+pub struct CachedData<'a> {
+  data: *const u8,
+  length: i32,
+  rejected: bool,
+  buffer_policy: BufferPolicy,
+  _phantom: PhantomData<&'a ()>,
+}
+
+impl<'a> Drop for CachedData<'a> {
+  fn drop(&mut self) {
+    unsafe {
+      v8__ScriptCompiler__CachedData__DELETE(self);
+    }
+  }
+}
+
+impl<'a> CachedData<'a> {
+  pub fn new(data: &'a [u8]) -> UniqueRef<Self> {
+    unsafe {
+      UniqueRef::from_raw(v8__ScriptCompiler__CachedData__NEW(
+        data.as_ptr(),
+        data.len() as i32,
+      ))
+    }
+  }
+}
+
+impl<'a> std::ops::Deref for CachedData<'a> {
+  type Target = [u8];
+  fn deref(&self) -> &Self::Target {
+    unsafe { std::slice::from_raw_parts(self.data, self.length as usize) }
+  }
+}
+
+#[repr(C)]
+#[derive(Debug)]
+enum BufferPolicy {
+  BufferNotOwned = 0,
+  BufferOwned,
+}
+
 impl Source {
-  // TODO(ry) cached_data
   pub fn new(source_string: Local<String>, origin: &ScriptOrigin) -> Self {
     let mut buf = MaybeUninit::<Self>::uninit();
     unsafe {
-      v8__ScriptCompiler__Source__CONSTRUCT(&mut buf, &*source_string, origin);
+      v8__ScriptCompiler__Source__CONSTRUCT(
+        &mut buf,
+        &*source_string,
+        origin,
+        std::ptr::null_mut(),
+      );
       buf.assume_init()
     }
   }
+
+  pub fn new_with_cached_data(
+    source_string: Local<String>,
+    origin: &ScriptOrigin,
+    cached_data: UniqueRef<CachedData>,
+  ) -> Self {
+    let mut buf = MaybeUninit::<Self>::uninit();
+    unsafe {
+      v8__ScriptCompiler__Source__CONSTRUCT(
+        &mut buf,
+        &*source_string,
+        origin,
+        cached_data.into_raw(),
```

Comment:
```suggestion
        cached_data.into_raw(), // Source constructor takes ownership.
```
Not immediately obvious from looking at just the code, I think.